### PR TITLE
Add default densification strategy (ADC: Adaptive Density Control)

### LINKED ISSUE #153: Add default densification strategy
Currently, this implementation supports only the mcmc densification strategy in mcmc.cpp/hpp. 

The goal of this issue is to implement the default densification.

The base implementation can be ported from: https://github.com/nerfstudio-project/gsplat/blob/main/gsplat/strategy/default.py

However, this can be only a starting point for later improvements.

Solving this issue should give a good understanding how the original densification strategy (adc) works.

### PR DESCRIPTION
**Pull request by Tomoya-Matsubara**
Closes #153 

The MCMC strategy implementation is refactored following [the implementation of `gsplat`](https://github.com/nerfstudio-project/gsplat/tree/2323de5905d5e90e035f792fe65bad0fedd413e7/gsplat/strategy).

### CODE REVIEW

**Comment by MrNeRF** [comment]
@Tomoya-Matsubara Thank you for this pr. It is quite big and I am busy most of the day today. I will review and test everything until Sunday evening if I don't get earlier to it.

File: src/default_strategy.cpp [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+        2 * (y * z + w * x),
+        1 - 2 * (x * x + y * y)};
+
+    auto R = torch::stack(R_components, -1);
+    auto shape = quats.sizes().vec();
+    shape.back() = 3;
+    shape.push_back(3);
+    return R.reshape(shape);
+}
+
```
[the remark this answers is not in the document]
Comment (Tomoya-Matsubara) [reply src/default_strategy.cpp]: This was debugging code. Removed.

**Comment by MrNeRF** [comment]
I just ran the benchmark script https://github.com/MrNeRF/gaussian-splatting-cuda/blob/master/scripts/benchmark_mipnerf360.sh and the first scene turns out quite low in quality. I stopped the benchmark here.

Interestingly, it becomes worse over time. I couldn't spot what might cause the issue on a first sight. Any idea?

-----------------
 Iteration      PSNR      SSIM     LPIPS    Time(s/img)     #Gaussians
---------------------------------------------------------------------------
      7000   23.3360    0.7545    0.2542         0.1801        2902341
     30000   22.3619    0.7381    0.2748         0.1619        2544362



image

**Comment by Tomoya-Matsubara** [comment]
## Hyperparameters

First, some hyperparameters need to be adjusted. An example of `optimization_params.json` looks like:

```diff
{
  ...
  "refine_every": 100,
  "start_refine": 500,
- "stop_refine": 25000,
+ "stop_refine": 15000,
  "grad_threshold": 0.0002,
  "sh_degree": 3,
- "opacity_reg": 0.01,
+ "opacity_reg": 0.0,
- "scale_reg": 0.01,
+ "scale_reg": 0.0,
- "init_opacity": 0.5,
+ "init_opacity": 0.1,
- "init_scaling": 0.1,
+ "init_scaling": 1.0,
  "max_cap": 1000000,
  ...
}
```

The updated values are taken from `gsplat`:

- [refine_stop_iter](https://github.com/nerfstudio-project/gsplat/blob/2323de5905d5e90e035f792fe65bad0fedd413e7/gsplat/strategy/default.py#L87)
- [opacity_reg, scale_reg](https://github.com/nerfstudio-project/gsplat/blob/2323de5905d5e90e035f792fe65bad0fedd413e7/examples/simple_trainer.py#L143-L146)
- [init_opacity, init_scale](https://github.com/nerfstudio-project/gsplat/blob/2323de5905d5e90e035f792fe65bad0fedd413e7/examples/simple_trainer.py#L216-L217)


## Divergences from the reference implementation

Second, there are several divergences from the `gsplat` and the original implementation. 

### `scene_scale`

18815e272ef21d926103be555fda5dc52f16048a changed the way to compute `scene_scale` as follows for more robustness, but this might affect the performance of the default strategy.

https://github.com/MrNeRF/gaussian-splatting-cuda/blob/ae5005d4bd80a2c99f9646f66ec03d51c4b06659/src/splat_data.cpp#L356-L357


### Initialization of `rotation`

`gsplat` [initializes `rotation` with random values](https://github.com/nerfstudio-project/gsplat/blob/73fad53c31ec4d6b088470715a63f432990493de/examples/simple_trainer.py#L254) while it starts as the identity quaternion in this repository:

https://github.com/MrNeRF/gaussian-splatting-cuda/blob/ae5005d4bd80a2c99f9646f66ec03d51c4b06659/src/splat_data.cpp#L381-L384

That said, I guess this difference does not change a lot.


### Initialization of `scaling`

In `gsplat`, `scaling` (i.e., `scales`) of a Gaussian is computed with:

$$
s \sqrt{\frac{1}{k} \sum_{i=1}^k d_i^2 }
$$

where $s$ is `init_scale` and $d_i$ is the distance to the $i$-th nearest Gaussian.

The implementation can be found [here](https://github.com/nerfstudio-project/gsplat/blob/2323de5905d5e90e035f792fe65bad0fedd413e7/examples/simple_trainer.py#L243-L246)

On the other hand, it is computed in this repository as:

$$
s \sqrt{\frac{1}{k} \sum_{i=1}^k d_i }
$$

In short, the initial `scaling` is likely to be smaller, which inhibits Gaussian growth under the default strategy. I'm not sure if this is intentional or a subtle bug, but it could contribute to the observed behavior.

https://github.com/MrNeRF/gaussian-splatting-cuda/blob/ae5005d4bd80a2c99f9646f66ec03d51c4b06659/src/splat_data.cpp#L373-L379

https://github.com/MrNeRF/gaussian-splatting-cuda/blob/ae5005d4bd80a2c99f9646f66ec03d51c4b06659/src/splat_data.cpp#L91-L104

**Comment by MrNeRF** [comment]
Ok, thank you. I will take a closer look and try to investigate!

**Comment by Tomoya-Matsubara** [comment]
I ran the full benchmark with the adjusted hyperparameters. Here are the results:

| Scene    | Iteration | PSNR          | SSIM         | LPIPS        | Num Gaussians |
| -------- | --------: | ------------: | -----------: | -----------: |-------------: |
| garden   | 7000     | 25.295799     | 0.758727     | 0.272304     | 3,304,898       |
| bicycle  | 7000     | 22.783640     | 0.607687     | 0.476105    | 3,052,815       |
| stump    | 7000     | 24.403950     | 0.716911     | 0.410456     | 3,464,276       |
| bonsai   | 7000     | 28.682089     | 0.915174     | 0.331841     | 941,320       |
| counter  | 7000     | 25.721880     | 0.874315     | 0.378319     | 699,188       |
| kitchen  | 7000     | 27.473415     | 0.887833     | 0.245382     | 954,480       |
| room     | 7000     | 27.730062     | 0.892442     | 0.401232     | 819,067       |


| Scene    | Iteration | PSNR          | SSIM         | LPIPS        | Num Gaussians |
| -------- | --------: | ------------: | -----------: | -----------: |-------------: |
| garden   | 30000     | 27.114607     | 0.835795     | 0.144486     | 4,937,304       |
| bicycle  | 30000     | 24.954300     | 0.731529     | 0.291470    | 5,684,053       |
| stump    | 30000     | 25.858850     | 0.767518     | 0.286427     | 4,647,623       |
| bonsai   | 30000     | 30.883152     | 0.927643     | 0.285990     | 1,120,498       |
| counter  | 30000     | 27.906197     | 0.906386     | 0.307963     | 886,049       |
| kitchen  | 30000     | 30.203331     | 0.921208     | 0.177520     | 1,129,135       |
| room     | 30000     | 30.404575     | 0.917164     | 0.336166     | 1,199,942       |


Compared to[ the `gsplat` report](https://docs.gsplat.studio/main/tests/eval.html), fewer Gaussians are generated, meaning there’s less duplication and splitting. This may be due to random variation and the differences in initialization.

**Comment by MrNeRF** [comment]
Verified! I am currently running the mcmc metrics to complete the measurements. Afterwards I will merge.

**Comment by MrNeRF** [comment]
That's what I get when I change

| Scene   | Iteration | PSNR      | SSIM     | LPIPS    | Num Gaussians |
|---------|-----------|-----------|----------|----------|----------------|
| garden  | 30000     | 26.656960 | 0.819221 | 0.187152 | 1,000,000      |
| bicycle | 30000     | 24.617630 | 0.722643 | 0.320389 | 1,000,000      |
| stump   | 30000     | 26.110689 | 0.777900 | 0.319444 | 1,000,000      |
| bonsai  | 30000     | 31.453180 | 0.947668 | 0.262173 | 1,000,000      |
| counter | 30000     | 28.404345 | 0.923728 | 0.260823 | 1,000,000      |
| kitchen | 30000     | 30.399292 | 0.929506 | 0.168228 | 1,000,000      |
| room    | 30000     | 31.863565 | 0.932434 | 0.287021 | 1,000,000      |
| mean    | 30000     | 28.500809 | 0.864729 | 0.257890 | 1,000,000      |

I only changed strategy in the parameters:
```
diff --git a/parameter/optimization_params.json b/parameter/optimization_params.json
index 6d5febf..010db4d 100644
--- a/parameter/optimization_params.json
+++ b/parameter/optimization_params.json
@@ -19,7 +19,7 @@
   "init_scaling": 0.1,
   "max_cap": 1000000,
   "render_mode": "RGB",
-  "strategy": "default",
+  "strategy": "mcmc",
   "eval_steps": [7000, 30000],
   "save_steps": [7000, 30000],
   "enable_eval": false,

```
I would expect to reproduce there metrics from master. Maybe there is a regression on master? I will rerun it there.

**Comment by MrNeRF** [comment]
So this is on master with mcmc
| scene   |   iteration |    psnr |     ssim |    lpips | num_gaussians   |
|:--------|------------:|--------:|---------:|---------:|:----------------|
| garden  |       30000 | 27.2099 | 0.837652 | 0.156712 | 1,000,000       |
| bicycle |       30000 | 25.3334 | 0.762709 | 0.253828 | 1,000,000       |
| stump   |       30000 | 26.7419 | 0.801107 | 0.262658 | 1,000,000       |
| bonsai  |       30000 | 32.4703 | 0.952533 | 0.249325 | 1,000,000       |
| counter |       30000 | 29.2731 | 0.930289 | 0.241491 | 1,000,000       |
| kitchen |       30000 | 31.5411 | 0.935336 | 0.154952 | 1,000,000       |
| room    |       30000 | 32.2489 | 0.937382 | 0.272846 | 1,000,000       |
| mean    |       30000 | 29.2598 | 0.879573 | 0.227402 |  1,000,000      |

If I change the strategy to mcmc, it should reproduce this result, isn't it? @Tomoya-Matsubara 
If I am not mistaken, the default works fine (with parameter adjustment) but mcmc doesn't?

**Comment by Tomoya-Matsubara** [comment]
The change includes refactoring of the implementation of MCMC to reduce redundancy, which may be causing some regression.

Since the default strategy seems to be working well, the refactoring part is suspicious. I'll check again!

**Comment by MrNeRF** [comment]
Ok, great. Let me know if you find something. Thank you!

**Comment by Tomoya-Matsubara** [comment]
I’ve identified the cause: the learning rate scheduler should apply only to the mean parameters in MCMC, but a refactor made it apply to all parameters. Fixed in 481dd96.

**Comment by MrNeRF** [comment]
Awesome! Thank you. I will test it.

**Comment by MrNeRF** [comment]
Can confirm! Mcmc results are 🔥

| scene   |   iteration |    psnr |     ssim |    lpips | num_gaussians   |
|:--------|------------:|--------:|---------:|---------:|:----------------|
| garden  |       30000 | 27.2549 | 0.838229 | 0.156464 | 1,000,000       |
| bicycle |       30000 | 25.3172 | 0.761948 | 0.255905 | 1,000,000       |
| stump   |       30000 | 26.5928 | 0.801281 | 0.262951 | 1,000,000       |
| bonsai  |       30000 | 32.4762 | 0.952130 | 0.249793 | 1,000,000       |
| counter |       30000 | 29.3200 | 0.930585 | 0.241641 | 1,000,000       |
| kitchen |       30000 | 31.6898 | 0.934465 | 0.156924 | 1,000,000       |
| room    |       30000 | 32.1135 | 0.937512 | 0.274300 | 1,000,000       |
| mean    |       30000 | 29.2521 | 0.879450 | 0.228283 |                 |


Approved! Thank you very much for your hard work @Tomoya-Matsubara. That was a very big task!

**Comment by dozeri83** [comment]
@MrNeRF @Tomoya-Matsubara 

getting build errors in windows because of strategy.cpp

tested in linux and it works.

@miriameng - do you also have it ?

error

```
C:\Users\User\dev\gaussian-splatting-cuda\include\core/strategy.hpp(22): error C2061: syntax error: identifier 'OptimizerParamState'
C:\Users\User\dev\gaussian-splatting-cuda\src\strategy.cpp(79): error C2061: syntax error: identifier 'OptimizerParamState'
C:\Users\User\dev\gaussian-splatting-cuda\src\strategy.cpp(111): error C2064: term does not evaluate to a function taking 2 arguments
C:\Users\User\dev\gaussian-splatting-cuda\src\strategy.cpp(112): error C3536: 'new_state': cannot be used before it is initialized
C:\Users\User\dev\gaussian-splatting-cuda\src\strategy.cpp(112): error C2679: binary '=': no operator found which takes a right-hand operand of type 'int' (or there is no acceptable conversion)
```


still did not figure out how to fix it

**Comment by MrNeRF** [comment]
Damn. I will put in some effort into a CI pipeline for Windows. That's certainly annoying as it is coming up on a daily basis. Won't be able to do it before 11 of August, but than I will prioritize this! I am deeply sorry about the inconvenience.

**Comment by dozeri83** [comment]
fix this in #234
